### PR TITLE
Fix Datagroup sharing by Virtual Servers

### PR DESCRIPTION
**Pull request by subbuv26**
Problem: Same Datagroup is being shared with all the Virtuals in CRD mode, which leads to security issue.

Solution: Create a dedicated Datagroup and iRule of each Virtual with the hosts it routes.

Signed-off-by: Subba Reddy Veeramreddy <subbareddyv.uoh@gmail.com>

### CODE REVIEW

**Comment by iam-veeramalla** [comment]
Hi @subbuv26 , I feel this is a very tedious change and should not be implemented unless we identified a considerable security issue. Can you please brief, what is the possible security incident identified?

**Comment by subbuv26** [comment]
@nandakishorepeddi please explain the issue.

File: pkg/crmanager/worker.go [hunk 1]
```diff
@@ -622,13 +619,19 @@ func (crMgr *CRManager) syncVirtualServers(
...
 
+		if isVSDeleted {
+			crMgr.handleVSDeleteForDataGroups(virtual, rsName)
+		}
+
 		// Delete rsCfg if no corresponding virtuals exist
 		// Delete rsCfg if it is HTTP rsCfg and the CR VirtualServer does not handle HTTPTraffic
 		if (len(virtuals) == 0) ||
 			(portStruct.protocol == "http" && !doesVSHandleHTTP(virtual)) {
-			crMgr.resources.deleteVirtualServer(rsName)
```
Comment (agupta49) [inline pkg/crmanager/worker.go:629]: @subbuv26 These changes are already incorporated into another PR.

File: pkg/crmanager/worker.go [hunk 2]
```diff
@@ -622,13 +619,19 @@ func (crMgr *CRManager) syncVirtualServers(
...
-			//TODO: what if last VirtualServer does not have cidr
-			crMgr.releaseIP(virtual.Spec.Cidr, virtual.Spec.Host)
+			crMgr.deleteVirtualServer(rsName)
+			if len(virtuals) == 0 && crMgr.ipamCli != nil {
```
Comment (agupta49) [inline pkg/crmanager/worker.go:631]: @subbuv26 These changes are already incorporated into another PR.

Comment (subbuv26) [reply pkg/crmanager/worker.go:631]: That is good. I will remove this line and update PR.

**Review by agupta49 — APPROVED** [review]

**Pull request merged**